### DISCUSSION TITLE
concourse and filesystems / resources

### DISCUSSION THREAD

**Discussion by maxnbk** [concourse/concourse#5495] [General]
New to concourse. Just got a hello-world pipeline running. Trying to figure out a good pattern of usage, here.

I'm in an environment where every host is assumed to have access to several network mounts. Many of the workloads I would like to run in concourse tasks would need to rely on reading the filesystems presented by these mounts. Not write-back, as I understand that to be a fairly "atomicity-destroying" concept and antithetical to concourse.

In our normal "build" pipeline (and by that I mean when we compile things), we run a command that, for example, mounts these filesystems to a docker container with the "-v" flag, typically as "ro", so that even baseline dependencies are accessible.

I'd like these filesystems mounted to the garden containers, but am unsure how to achieve it, declaratively. For example, I can't seem to find any reference in terms of how to handle "image_resource" of type "docker" in such a way that would allow any sort of usage like that.

By the same token, these filesystems are.... Large. Many terabytes. Pushing this stuff into git or something else to track as a resource in that way will simply not work. I'm internal to a corporate network, so pushing our entire filesystem to s3 is also not in the cards.

I don't necessarily need jobs to trigger on changes to these filesystems, I just need them as a baseline assumption of the container, even if I have to declare it in some way.

Is there any facility for this sort of use-case? Am I missing something obvious?

There are some other assumptions I can make, at the time that I would "set" a pipeline, which would be things like, exactly which directories would need polling, if a resource type exists that could express "does this directory `ls` differently than it did last time I ran", so if treating individual directories as resources is possible, that does make a degree of reactivity to changes possible, but does not change the need to be able to see those directories via the filesystem network mounts in the first place. Other possibilities would be a specific git check acting as a proxy for the fact that a directory changed, which could be very useful too.

A solution to this will unlock a huge number of possibilities for me to play with, so I hope I'm at least barking up something resembling a tree, here. Thank you for anything I can go off of here.

**Comment by vito** [comment]
There's some related discussion around this here: https://github.com/concourse/concourse/issues/2353

Short story is there is currently no way to do this, but it's come up before. Supporting read-only mounts from the worker seems reasonable to me, but it's not currently implemented. Ideally this would be something configured on the worker side.

**Reply by maxnbk** [reply]
So essentially, the only way to get files into a container is via something like blob storage, git resource, etc? Are there any options I've overlooked that I could hijack? Neither of those is an option in my case (well, the git part is, but only one small component)

**Reply by vito** [reply]
Yep. Concourse is pretty particular about this; it tries really hard to keep you from being tightly coupled to state on your workers and instead wants you to push as much as possible into Resources as the "source of truth." Use cases like yours are valid counterpoints, they just weren't motivators for the original design.

Hmm I can't think of any easy workarounds. I think you'd need to change some code either way. There isn't really a code path that allows users to inject arbitrary mount points (just saying that out loud it would probably be a massive security hole, haha). (There isn't an operator-level configuration for it either.)

If you're willing to hack some code together, you could probably add a flag to `concourse web` that takes a mapping from "worker tag" => "volume mounts", and modify the `atc/worker/` package  (guessing here) to append the given mount configuration for any containers it creates on a worker matching the given tag. Preferably only Task containers.

This would probably be the quickest path to becoming unblocked, but it's not really a design we'd want in the long term, FYI. We try to minimize cluster-wide configuration like that and imagine `concourse web` is used by many different teams that don't know about each other.

A more long-term option would be to implement support for this for the `containerd` worker backend, which is an upcoming replacement for the current `guardian` backend. A flag for volume mounts could be configured on the worker side and injected into the OCI specs that it generates. This is possible now because we control the backend code, but note that the `containerd` workers are still a work in progress and there may be dragons.

The most expensive option would be to implement it for the current workers, which would involve adding configuration to the workers which then gets sent back to the ATC (similar to [`CertsPath`](https://github.com/concourse/concourse/blob/d6ac1346bdf3bc74878caabc8a8bf0ad42b67121/atc/worker.go#L13), saved in the database, and then used for any Task containers which run on the worker. This would be a bit more work but I'm happy to explain more if you're interested.

**Reply by twitchyliquid64** [reply]
Could this instead be an option in guardian worker configuration? That way it only needs to be plumbed from config --> atc/.../mounts.go and not touch the database or anything like that.

**Reply by twitchyliquid64** [reply]
Reading through the source this does not appear to work in the way i expect ... Could you point me to the bit in the code that glues commands coming in from the web thingy, to some struct that specifies how a container is to be run/setup (I'm assuming a backend-agnostic representation exists somewhere?)

**Reply by twitchyliquid64** [reply]
Is `atc/exec/TaskStep` the abstract representation generated by the web thingy and then shipped to workers? and then some mystery glue yeets it to the container backend and yoinks back outputs?

**Reply by jamieklassen** [reply]
`atc/worker/Client` is the mystery glue that yeets and yoinks:

https://github.com/concourse/concourse/blob/82a76b64928ed8d1b0a6500d4bf958422d09e675/atc/worker/client.go#L30-L96

**Reply by twitchyliquid64** [reply]
Perfect, so I think I could just modify `FindOrCreateContainer` to pass through worker-specific mounts from ~local (worker)~ web/atc configuration, and add that to the container spec just before it ends up at the backend. Thoughts?

**Reply by maxnbk** [reply]
such an implementation (whether passed in some way at worker runtime or somehow via a pipeline configuration) would be plenty sufficient for our usage. I can't speak to the correctness regarding concourse "best practices", but at the moment, without this, I'm stuck in advancing concourse in our world.

**Reply by maxnbk** [reply]
Circling back around to this wistfully - Has the landscape changed here in the past several feature versions? I recall writing this around v6, and I guess we're at 6.4.1 now. Willing to possibly take a crack at (or in collab with someone) on this, given some updates or additional direction. Did you try anything, @twitchyliquid64 ?

**Reply by vito** [reply]
@maxnbk A PR was opened but not completed: https://github.com/concourse/concourse/pull/5684 - maybe it could be continued from there?
